File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2019/Y19D17.kt
package org.gristle.pdxpuzzles.advent.y2019

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.advent.y2019.Intcode.IntCode
import org.gristle.pdxpuzzles.utilities.enums.Nsew
import org.gristle.pdxpuzzles.utilities.objects.Coord
import org.gristle.pdxpuzzles.utilities.objects.Grid
import org.gristle.pdxpuzzles.utilities.objects.toMutableGrid
import java.util.*

class Y19D17(input: String) : Day {
    private val initialState = input.split(',').map { it.toLong() }

    fun solve(): Pair<Int, Long> {

        val toDroid: Deque<Long> = LinkedList()
        val toComp: Deque<Long> = LinkedList()
        val intCode = IntCode("A", initialState, null, toComp, toDroid)
        intCode.run()
        val width = toDroid.indexOfFirst { it == 10L }
        val grid = toDroid.filter { it != 10L }.toMutableGrid(width)
        val intersections = grid.mapIndexedNotNull { index, l ->
            if (l != 35L || grid.getNeighbors(index).any { it != 35L }) {
                null
            } else {
                val pos = grid.coordOf(index)
                pos.x * pos.y
            }
        }
        val p1 = intersections.sum()

        // Part 2
        println(grid.representation { it.toInt().toChar() })
        var coord = grid.coordOfElement(94L)
        var dir = Nsew.NORTH
        var counter = 0
        val pathBuilder = StringBuilder()
        while (true) {
            if (cromulent(grid, coord, dir)) {
                counter++
                coord = dir.forward(coord)
            } else if (cromulent(grid, coord, dir.left())) {
                dir = dir.left()
                if (counter != 0) {
                    pathBuilder.append(counter)
                    pathBuilder.append(',')
                    counter = 0
                }
                pathBuilder.append("L,")
            } else if (cromulent(grid, coord, dir.right())) {
                dir = dir.right()
                if (counter != 0) {
                    pathBuilder.append(counter)
                    pathBuilder.append(',')
                    counter = 0
                }
                pathBuilder.append("R,")
            } else {
                pathBuilder.append(counter)
                pathBuilder.append(',')
                break
            }
        }
        val path = pathBuilder.toString()
        val matches = Regex("""[LR],\d+,""").findAll(path).toList()
        val (formSeq, forms) = getCommands(path, matches)
        val video = listOf('n', '\n')
        val commands = (formSeq + forms[0] + forms[1] + forms[2] + video).map { it.code.toLong() }
        toComp.addAll(commands)
        val intCodeB = IntCode("B", listOf(2L) + initialState.drop(1), null, toComp, toDroid)
        toDroid.clear()
        intCodeB.run()
        return p1 to toDroid.last()
    }

    private fun getCommands(path: String, matches: List<MatchResult>): Pair<String, List<String>> {
        val commands = mutableListOf<StringBuilder>()
        splitCommands(path, matches, 0, commands, 3)
        val formSequence = StringBuilder()
        var cursor = 0
        while (cursor < path.length) {
            val position = commands.indexOfFirst { s -> path.drop(cursor).startsWith(s) }
            val c = when (position) {
                0 -> 'A'
                1 -> 'B'
                2 -> 'C'
                else -> throw IllegalStateException()
            }
            formSequence.append(c)
            formSequence.append(',')
            cursor += commands[position].length
        }
        for (command in commands) {
            command.setLength(command.length - 1)
            command.append('\n')
        }
        formSequence.setLength(formSequence.length - 1)
        formSequence.append('\n')
        return formSequence.toString() to commands.map { it.toString() }
    }

    private fun splitCommands(
        path: String,
        matches: List<MatchResult>,
        matchIndex: Int,
        commands: MutableList<StringBuilder>,
        level: Int
    ): Boolean {
        if (level == 0 && matches.size == 0) {
            return true
        }
        var start = matches[matchIndex].range.first
        outer@while (true) {
            for (command in commands) {
                if (path.drop(start).startsWith(command.toString())) {
                    start += command.length
                    continue@outer
                }
            }
            break
        }
        if (level == 0) {
            return start >= path.length
        }
        for ((index, match) in matches
            .drop(matchIndex)
            .withIndex()
            .filter { (_, m) -> m.range.last <= start + 20 }
            .asReversed())
        {
            val sb = StringBuilder()
            sb.append(path.substring(start .. match.range.last))
            commands.add(sb)
            if (splitCommands(path, matches, matchIndex + index + 1, commands, level - 1)) {
                return true
            } else {
                commands.removeLast()
            }
        }
        return false
    }

    private fun cromulent(grid: Grid<Long>, coord: Coord, dir: Nsew): Boolean {
        val prospect = dir.forward(coord)
        return (prospect.x in 0 until grid.width && prospect.y in 0 until grid.height) && grid[prospect] == 35L
    }

    private val solution = solve()

    override fun part1() = solution.first

    override fun part2() = solution.second
}

fun main() = Day.runDay(Y19D17::class)

//    Class creation: 56ms
//    Part 1: 10632 (1ms)
//    Part 2: 1356191 (1ms)
//    Total time: 60ms

File: rust/advent/src/bin/2019/y19d17.rs
use advent::utilities::{get_input::get_input, intcode::IntCode};
use lazy_regex::regex;
use lazy_regex::regex::Match;
use utilities::{
    enums::cardinals::Cardinal,
    parsing::get_numbers::ContainsNumbers,
    structs::{
        coord::Coord2U,
        grid::{Grid2, GridIterator},
        stopwatch::{ReportDuration, Stopwatch},
    },
};

type Input = (Vec<i64>, Grid2<i64>);
type Output = usize;
type Pos = Coord2U;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(19, 17).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    let mut code: Vec<_> = input.get_numbers().collect();
    let mut ic = IntCode::new(&code);
    let (_, output) = ic.run_while_able();
    let width = output.iter().position(|&it| it == 10).unwrap();
    let grid = output
        .into_iter()
        .filter(|&it| it != 10)
        .try_collect_grid(width)
        .unwrap();
    code[0] = 2;
    (code, grid)
}

fn part1(input: &Input) -> Output {
    let (_, grid) = input;
    grid.iter()
        .enumerate()
        .filter_map(|(index, &l)| {
            if l != 35
                || grid
                    .adjacent(index, false)
                    .unwrap()
                    .any(|adj| *adj.value != 35)
            {
                None
            } else {
                let pos = grid.coord_of(index).unwrap();
                Some(pos.x() * pos.y())
            }
        })
        .sum()
}

fn part2(input: &Input) -> Output {
    let (code, grid) = input;
    let index = grid.iter().position(|&it| it == 94).unwrap();
    let mut pos = grid.coord_of(index).unwrap();
    let mut dir = Cardinal::North;
    let mut counter = 0;
    let mut path = String::new();
    let l_str = "L,";
    let r_str = "R,";
    loop {
        if let Some(prospect) = cromulent(grid, pos, dir) {
            counter += 1;
            pos = prospect;
        } else if let Some(_) = cromulent(grid, pos, dir.left()) {
            dir = dir.left();
            if counter != 0 {
                path.push_str(counter.to_string().as_str());
                path.push(',');
                counter = 0;
            }
            path.push_str(l_str);
        } else if let Some(_) = cromulent(grid, pos, dir.right()) {
            dir = dir.right();
            if counter != 0 {
                path.push_str(counter.to_string().as_str());
                path.push(',');
                counter = 0;
            }
            path.push_str(r_str);
        } else {
            path.push_str(counter.to_string().as_str());
            path.push(',');
            break;
        }
    }
    let matches: Vec<Match> = regex!(r"(?:L|R),\d+,").find_iter(&path).collect();
    let (form_seq, forms) = get_commands(&path, &matches);
    let video = ['n', '\n'];
    let to_ic: Vec<i64> = form_seq.chars()
        .chain(forms.iter().flat_map(|form| form.chars()))
        .chain(video.into_iter())
        .map(|c| c as i64)
        .collect();
    let mut ic = IntCode::new(&code);
    ic.input_slice(&to_ic);
    let (_, output) = ic.run_while_able();
    output[output.len() - 1] as usize
}

fn cromulent(grid: &Grid2<i64>, pos: Pos, dir: Cardinal) -> Option<Pos> {
    let Some(prospect) = pos.move_direction(dir, 1) else {
        return None;
    };
    if *grid.get(prospect).unwrap_or(&0) == 35 {
        Some(prospect)
    } else {
        None
    }
}

fn get_commands(path: &str, matches: &[Match]) -> (String, Vec<String>) {
    let mut commands: Vec<String> = Vec::new();
    split_commands(path, matches, &mut commands, 3);
    let mut form_sequence = String::new();
    let mut cursor = 0;
    while cursor < path.len() {
        let position = commands.iter()
            .position(|s| path[cursor..].starts_with(s))
            .unwrap();
        let c = match position {
            0 => 'A',
            1 => 'B',
            2 => 'C',
            _ => unreachable!(),
        };
        form_sequence.push(c);
        form_sequence.push(',');
        cursor += commands[position].len();
    }
    for command in commands.iter_mut() {
        command.pop();
        command.push('\n');
    }
    form_sequence.pop();
    form_sequence.push('\n');
    (form_sequence, commands)
}

fn split_commands(
    path: &str,
    matches: &[Match],
    commands: &mut Vec<String>,
    level: usize
) -> bool {
    // base case
    if level == 0 && matches.len() == 0 {
        return true;
    }
    let mut start = matches[0].start();
    'outer: loop {
        for command in commands.iter() {
            if path[start..].starts_with(command) {
                start += command.len();
                continue 'outer;
            }
        }
        break;
    }
    if level == 0 {
        return start >= path.len();
    }
    for (index, mat) in matches.iter()
        .enumerate()
        .filter(|(_, m)| m.end() <= start + 20)
        .rev()
    {
        commands.push(path[start..mat.end()].to_string());
        if split_commands(path, &matches[index + 1..], commands, level - 1) {
            return true;
        } else {
            commands.pop();
        }
    }
    return false
}

#[test]
fn default() {
    let input = get_input(19, 17).unwrap();
    let input = parse_input(&input);
    assert_eq!(10632, part1(&input));
    // assert_eq!(1356191, part2(&input));
}

// Input parsed (490μs)
// 1. 10632 (28μs)
// 2. 1356191 (1ms)
// Total: 1ms